### PR TITLE
Fix resource method related compiler crashes

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
- Stop resource function being included in object type inclusions.
- Fix CCE in node transformer.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31611
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31636
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31906

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal [hunk 1]
```diff
@@ -23,3 +23,23 @@ type SType service object {
...
 }
+
+service class ROne {
+    resource function get foo() returns string {
+        return "foo";
+    }
+}
+
+// Not having a implementation for "get foo" resource method is expected as resource method is not part of the obj type.
+service class RTwo {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:34]: Shall we also add an assignability tests with inclusions?

```ballerina
service class Foo {
    resource function get greeting() returns string => "hello";
}

service class Bar {
    *Foo;

    function fn() {

    }
}

service class Baz {
    function fn() {
        
    }
}

public function main() {
    Bar x = new Baz(); // allowed now
}
```

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:34]: Came across https://github.com/ballerina-platform/ballerina-lang/issues/31906 btw.

**Review by MaryamZi — COMMENTED** [review]
I guess we'll have to send an update on this change on Slack.

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal (hunk 1 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:34]: Added with https://github.com/ballerina-platform/ballerina-lang/pull/31761/commits/ada5ceb039c3ea0e94ad9b091ceab82b76df7a64

File: tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal [hunk 2]
```diff
@@ -23,3 +23,41 @@ type SType service object {
...
+
+service class DoDone {
+    resource function 'do f() returns int => 0;
+
+    resource function done f() returns int => 0;
+
+    public function 'do() returns int => 0;
+}
+
+service class Do {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:56]: Shall we add separate tests for the fix done for https://github.com/ballerina-platform/ballerina-lang/issues/31906 with mismatched function counts? May also have to add a runtime test if we don't have one already, right (IIRC it worked as expected)?

Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:56]: You want me to move negative test's bal code to a separate file or to a separate test class?
The runtime typechecking approach is drastically different from the compiler's short circuiting approach

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:56]: No no, separate tests for cases like (without inclusion)

```ballerina
service class Foo {
    resource function get greeting() returns string => "hello";

    remote function hello() {
        
    }
}

service class Bar {
    remote function hello() {
        
    }
}

function positive() {
    Foo f = new Bar(); // allowed now
}

service class Baz {

}

function negative() {
    Foo f = new Baz(); // still not allowed
}
```

Re: runtime tests, just realized the runtime methods are modeled separately. Came across https://github.com/ballerina-platform/ballerina-lang/issues/31942 though. I guess we can add runtime tests when fixing that.

Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/klass/service_type_resource_method_decl_neg.bal:56]: fixed in https://github.com/ballerina-platform/ballerina-lang/pull/31761/commits/01c4a1d16ab1975e72e4f587af89037fa089d351

**Comment by MaryamZi** [comment]
LGTM. Shall we send an update on Slack before merging this?

**Comment by rdhananjaya** [comment]
> LGTM. Shall we send an update on Slack before merging this?

Sent an update with https://ballerina-platform.slack.com/archives/C9WMA6QDC/p1627567586115900

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**